### PR TITLE
opt: use join filters to imply IS NOT NULL partial index predicates

### LINKED ISSUE #57841: Partial index not being used for foreign key check when deleting a row
**Describe the problem**

I work on the Ion team at Tesla.

We have a table (observations) that has a foreign key to another table (carriers).  When deleting a record from carriers, a full table scan is performed on observations for records containing that key, instead of using the partial index.  

**To Reproduce**

Observations table with the following schema

``` sql
CREATE TABLE public.observations (
	id UUID NOT NULL DEFAULT gen_random_uuid(),
	created_at TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
	observation_group_id UUID NULL,
	component_id STRING NULL,
	carrier_id STRING NULL,
	location_id UUID NULL,
	process_step_id UUID NULL,
	process_task_id UUID NULL,
	subject_path STRING NOT NULL,
	observer_path STRING NULL,
	categories INT2[] NULL,
	observed_at TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
	secondary_observable_id UUID NULL,
	secondary_observable_kind INT2 NULL,
	CONSTRAINT "primary" PRIMARY KEY (id ASC),
	CONSTRAINT fk_observation_group FOREIGN KEY (observation_group_id) REFERENCES public.observation_groups(id),
	CONSTRAINT fk_component FOREIGN KEY (component_id) REFERENCES public.components(id),
	CONSTRAINT fk_carrier FOREIGN KEY (carrier_id) REFERENCES public.carriers(id),
	CONSTRAINT fk_location FOREIGN KEY (location_id) REFERENCES public.locations(id),
	CONSTRAINT fk_process_step FOREIGN KEY (process_step_id) REFERENCES public.process_steps(id),
	CONSTRAINT fk_process_task FOREIGN KEY (process_task_id) REFERENCES public.process_tasks(id),
	CONSTRAINT fk_secondary_observable FOREIGN KEY (secondary_observable_id, secondary_observable_kind) REFERENCES public.secondary_observables(id, kind),
	INDEX observations_group_idx (observation_group_id ASC) WHERE observation_group_id IS NOT NULL,
	INDEX observations_location_observed_idx (location_id ASC, observed_at DESC) WHERE location_id IS NOT NULL,
	INDEX observations_step_observed_idx (process_step_id ASC, observed_at DESC) WHERE process_step_id IS NOT NULL,
	INDEX observations_task_observed_idx (process_task_id ASC, observed_at DESC) WHERE process_task_id IS NOT NULL,
	INDEX observations_component_observed_idx (component_id ASC, observed_at DESC) WHERE component_id IS NOT NULL,
	INDEX observations_component_subject_observed_idx (component_id ASC, subject_path ASC, observed_at DESC) WHERE component_id IS NOT NULL,
	INDEX observations_carrier_observed_idx (carrier_id ASC, observed_at DESC) WHERE carrier_id IS NOT NULL,
	INDEX observations_carrier_subject_observed_idx (carrier_id ASC, subject_path ASC, observed_at DESC) WHERE carrier_id IS NOT NULL,
	INDEX observations_secondary_observable_observed_idx (secondary_observable_id ASC, observed_at DESC) WHERE secondary_observable_id IS NOT NULL,
	INDEX observations_secondary_observable_subject_observed_idx (secondary_observable_id ASC, subject_path ASC, observed_at DESC) WHERE secondary_observable_id IS NOT NULL,
	INDEX observations_secondary_observable_kind_idx (secondary_observable_id ASC, secondary_observable_kind ASC) WHERE secondary_observable_id IS NOT NULL,
	FAMILY "primary" (id, created_at, observation_group_id, component_id, carrier_id, location_id, process_step_id, process_task_id, subject_path, observer_path, categories, observed_at, secondary_observable_id, secondary_observable_kind),
	CONSTRAINT belongs_to CHECK ((((component_id IS NOT NULL)::INT8 + (carrier_id IS NOT NULL)::INT8) + (secondary_observable_id IS NOT NULL)::INT8) = 1:::INT8)
)
```
Note the foreign key constraint on `carrier_id` and the partial index 
```sql
INDEX observations_carrier_observed_idx (carrier_id ASC, observed_at DESC) WHERE carrier_id IS NOT NULL,
```

carriers table with the following schema:

``` sql
CREATE TABLE public.carriers (
	id STRING NOT NULL,
	site_id STRING NOT NULL,
	scope STRING NOT NULL,
	serial STRING NOT NULL,
	created_at TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
	retired_at TIMESTAMP NULL,
	kind INT2 NOT NULL,
	size INT2[] NULL,
	CONSTRAINT "primary" PRIMARY KEY (id ASC),
	CONSTRAINT fk_site FOREIGN KEY (site_id) REFERENCES public.sites(id),
	UNIQUE INDEX carriers_site_scope_serial_idx (site_id ASC, scope ASC, serial ASC),
	FAMILY "primary" (id, site_id, scope, serial, created_at, retired_at, kind, size)
)
```

When running a delete on carriers we get the following explain:

``` sql
explain delete from carriers where id = 'test';
```
```
	distribution	local
	vectorized	false
root		
 ├── delete		
 │    │	from	carriers
 │    └── buffer		
 │         │	label	buffer 1
 │         └── scan		
 │	estimated row count	1
 │	table	carriers@primary
 │	spans	[/'test' - /'test']
 ├── fk-cascade		
 │	fk	fk_carrier
 │	input	buffer 1
 └── fk-check		
      └── error if rows		
           └── hash join (semi)		
                │	equality	(id) = (carrier_id)
                │	left cols are key	
                ├── scan buffer		
                │	label	buffer 1
                └── scan		
	estimated row count	24283827
	table	observations@primary
	spans	FULL SCAN
```

**Expected behavior**
I expected an index scan on the observations table for just the id in question

**Environment:**
 - CockroachDB version: v20.2.0
 - Server OS: Linux (not sure what distro)
 - Client app: Postico

**Additional context**
This is currently preventing deletes from the carriers table as the table scan takes too long to complete and times out.

### PR DESCRIPTION
**Pull request by mgartner**
#### opt: use join filters to imply IS NOT NULL partial index predicates

This commit updates the reject null normalization rules so that:

1. `RejectNullsUnderJoinRight` matches `SemiJoin` and `AntiJoin`.
2. A `Scan` with a `col IS NOT NULL` predicate requests null rejection
   of `col`.

In combination, these two changes allow partial indexes with
`IS NOT NULL` predicates to be used in more cases.

As an example, they can be used in JOINs where the ON condition
implicitly implies the predicate:

    CREATE TABLE a (a INT PRIMARY KEY)
    CREATE TABLE b (b INT, INDEX (b) WHERE b IS NOT NULL)

    SELECT * FROM a JOIN b ON a = b

As a another example, partial indexes with `IS NOT NULL` predicates can
be used to satisfy foreign key checks.

    CREATE TABLE parent (id INT PRIMARY KEY)

    CREATE TABLE child (
      id INT PRIMARY KEY,
      p_id INT,
      CONSTRAINT fk FOREIGN KEY (p_id) REFERENCES parent(id),
      INDEX (p_id) WHERE p_id IS NOT NULL
    )

    DELETE FROM p WHERE id = 1

The `DELETE` requires a foreign key check to ensure that no existing
rows in `child` have a matching `p_id`. Prior to this commit, the check
performed a full table scan rather than using the partial index because
its predicate was not implied. By pushing down the null-rejecting filter
derived from the check's `child.p_id = parent.id` filter, the partial
index can be used.

Fixes #57841

Release justification: This improves performance for joins and foreign
key checks on tables with partial indexes. Partial indexes are a new
feature introduced in 20.2.

Release note (performance improvement): Partial indexes with IS NOT NULL
predicates can be used in cases where JOIN filters implicitly imply the
predicate. This results in more efficient query plans for JOINs and
foreign checks.

### CODE REVIEW

File: pkg/sql/opt/testutils/opttester/opt_tester.go [hunk 1]
```diff
@@ -661,14 +663,14 @@ func fillInLazyProps(e opt.Expr) {
 		norm.DerivePruneCols(rel)
 
 		// Derive columns that are candidates for null rejection.
-		norm.DeriveRejectNullCols(rel)
+		norm.DeriveRejectNullCols(md, rel)
```
Comment (mgartner) [inline pkg/sql/opt/testutils/opttester/opt_tester.go:666]: I'm open to better ideas on how to pipe the metadata here, if you folks have any.

**Review by RaduBerinde — COMMENTED** [review]
Thanks for working on this! :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPBCO-DAgQdXAXVCndk:b7nhl9q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBCal77WgeFwZL6YjU:-MPBAnP53_s1-3ituetR:beqf6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/norm/reject_nulls_funcs.go#L333)):*
> ```Go
> 
> 	var rejectNullCols opt.ColSet
> 	for _, pred := range md.TableMeta(scan.Table).PartialIndexPredicates {
> ```

We should be able to use `in.Memo().Metadata()` here instead of plumbing.
___
*[pkg/sql/opt/testutils/opttester/opt_tester.go, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBAJgFcc2PL2zmLOec-r1-666:-MPBBhzD3wJg1rFNCdW_:b-67dcg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/testutils/opttester/opt_tester.go#L666)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm open to better ideas on how to pipe the metadata here, if you folks have any.
</blockquote></details>

I don't see a better way..
___
*[pkg/sql/opt/testutils/opttester/opt_tester.go, line 1568 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBBs6L9UcjzazUmW0y:-MPBBs6L9UcjzazUmW0z:bkbjcua) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/testutils/opttester/opt_tester.go#L1568)):*
> ```Go
> }
> 
> func (ot *OptTester) optimizeExpr(o *xform.Optimizer) (opt.Expr, *opt.Metadata, error) {
> ```

[nit] This is a bit odd, since the caller is passing in the Optimizer (with which the Metadata is associated). I'd leave this unchanged and use `o.Factory().Metadata()` at the call site.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBCal77WgeFwZL6YjU:-MPBSRwG8q9dtzjfaAZC:b-t8zwew) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/norm/reject_nulls_funcs.go#L333)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should be able to use `in.Memo().Metadata()` here instead of plumbing.
</blockquote></details>

Lol this is the "better way"! Now we don't need to make any opttester changes! :tada:

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> Reviewed 3 of 6 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPCKdPC4owpLTVVvWSA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic, line 971 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPCK_99DFmmVG1wNBvz:-MPCK_99DFmmVG1wNBw-:bz0auhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacf852e6107090285c8bce789d94fee32497f5/pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic#L971)):*
> ```
> Del /Table/57/1/20/0
> 
> # ---------------------------------------------------------
> ```

I think this should go into a `partial_index` file since it doesn't involve traces.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic, line 971 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPCK_99DFmmVG1wNBvz:-MPFtyHE2XwBoDwXJMWE:b-7s49rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacf852e6107090285c8bce789d94fee32497f5/pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic#L971)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this should go into a `partial_index` file since it doesn't involve traces.
</blockquote></details>

Done. Moved over the EXPLAIN tests too in a separate commit.

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPGa6h7EWDG3gW41XV5:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPH3Xmx3CHZ5J5RnB6j:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

**Comment by mgartner** [comment]
TFTR!

bors r=RaduBerinde

**Pull request merged**